src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/ThemeManager.cs

```diff
@@ -1,278 +1,477 @@
-    ///   If windows is not provided, apply the theme to all windows in the application.
-    /// </summary>
-    /// <param name="window"></param>
-    /// <param name="forceUpdate"></param>
-    internal static void ApplySystemTheme(IEnumerable windows = null, bool forceUpdate = false)
+    internal static void OnWindowThemeChanged(Window window, ThemeMode oldThemeMode, ThemeMode newThemeMode)
     {
-        if(windows == null)
+        if(newThemeMode == ThemeMode.None)
         {
-            // If windows is not provided, apply the theme to all windows in the application.
-            windows = Application.Current?.Windows;
-            
-            if(windows == null)
+            if(newThemeMode != oldThemeMode)
             {
-                return;
+                RemoveFluentFromWindow(window);
             }
+            return;
         }
 
-        string systemTheme = GetSystemTheme();
-        bool useLightMode = IsSystemThemeLight();
-        Color systemAccentColor = AccentColorHelper.SystemAccentColor;
-        ApplyTheme(windows , systemTheme, useLightMode, systemAccentColor, forceUpdate);
+        ApplyFluentOnWindow(window);
     }
 
-    /// <summary>
-    ///  Apply the requested theme and color mode to the windows.
-    ///  Checks if any update is needed before applying the changes.
-    /// </summary>
-    /// <param name="windows"></param>
-    /// <param name="requestedTheme"></param>
-    /// <param name="requestedUseLightMode"></param>
-    /// <param name="requestedAccentColor"></param>
-    /// <param name="forceUpdate"></param>
-    private static void ApplyTheme(
-        IEnumerable windows, 
-        string requestedTheme, 
-        bool requestedUseLightMode,
-        Color requestedAccentColor, 
-        bool forceUpdate = false)
+    internal static bool SyncThemeModeAndResources()
     {
-        if(forceUpdate || 
-                requestedTheme != _currentApplicationTheme || 
-                requestedUseLightMode != _currentUseLightMode ||
-                requestedAccentColor != _currentSystemAccentColor)
+        if(DeferSyncingThemeModeAndResources) return true;
+
+       ThemeMode themeMode = GetThemeModeFromResourceDictionary(Application.Current.Resources);
+
+        if(Application.Current.ThemeMode != themeMode)
         {
+            Application.Current.ThemeMode = themeMode;
+            return themeMode == ThemeMode.None ? false : true;
+        }
+        return false;
+    }
 
-            Uri dictionaryUri = GetFluentThemeResourceUri(requestedTheme, requestedUseLightMode);
-            AddOrUpdateThemeResources(dictionaryUri);
+    internal static void SyncDeferredThemeModeAndResources()
+    {
+        if(Application.Current == null) return;
+
+        ThemeMode themeMode = Application.Current.ThemeMode;
+
+        bool resyncThemeMode = false;
+        int index = FindLastFluentThemeResourceDictionaryIndex(Application.Current.Resources);
```

Comment:
I think the convention .NET uses would make it `LastIndexOfFluentThemeDictionary`, but probably doesn't matter here. Either way, it doesn't look like you actually care about the `index` value below, only whether it exists or not. And if it does, you look for the index again in `GetThemeModeFromResourceDictionary`. Why don't you do something like `bool TryGetThemeModeFromResourceDictionary(resources, out themeMode)` instead?

Reply:
Resolved this, after modification it's not that we are not using the fetched index, so I haven't changed to TryGet...., however I have followed the LastIndexOf... convention